VisualizationBase/src/items/ViewItem.cpp

```diff
@@ -102,38 +116,71 @@ const QPoint ViewItem::positionOfNode(Model::Node *node) const
 	return QPoint(-1, -1);
 }
 
-QVector<QVector<Model::Node*>> ViewItem::nodesGetter()
+const QPoint ViewItem::positionOfItem(Item *item) const
 {
-	return nodes_;
+	auto vref = DCast<VNodeReference>(item);
+	if (vref)
+		return positionOfNode(vref->node());
+	else return QPoint(-1, -1);
 }
 
-void ViewItem::ensureColumnExists(int column)
+Model::Node* ViewItem::nodeAt(int column, int row)
 {
-	if (nodes_.size() <= column)
-		nodes_.resize(column + 1);
+	if (column < 0 || column >= nodes_.size())
+		return nullptr;
+	if (row < 0 || row >= nodes_[column].size())
+		return nullptr;
+	return nodes_[column][row];
 }
 
-QString ViewItem::debugNodes()
+void ViewItem::addEmptyNode(int column, int row, int desiredYPosOfNextItem)
 {
-	//Prints the node grid as a debug string
-	QString result;
-	int maxHeight = 0;
-	for (int i = 0; i < nodes_.size(); i++)
-		if (nodes_.at(i).size() > maxHeight)
-			maxHeight = nodes_.at(i).size();
+	emptyNodesToBuild_.append(EmptyNodeParameters(column, row, desiredYPosOfNextItem));
+	setUpdateNeeded(StandardUpdate);
+}
 
-	for (int row = 0; row < maxHeight; row++)
+void ViewItem::updateGeometry(int availableWidth, int availableHeight)
+{
+	Super::updateGeometry(availableWidth, availableHeight);
+	while (!emptyNodesToBuild_.isEmpty())
```

Comment:
This code right now runs only once when we initially create a empty node, but in general empty nodes should adjust on every redraw based on the contents of nodes around them.

I previously said that you'll need to do these computations in updateChildren method of VViewItemNode. If you are careful, you could also do it here, in ViewItem::updateGeometry once for all nodes. This might be more convenient, but make sure to keep both an EmptyNode's item and node syncrhonized when you change the size.